pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,348 @@
+	"github.com/cockroachdb/redact"
+)
+
+// traceValue represents the trace information for a single registration.
+type traceValue struct {
+	traced kvserverpb.TracedEntry
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+	// baseCtx is the underlying proposal buffer which we additionally log to.
+	// We can't store the underlying context directly because it can be modified
+	// by the proposal buffer once it is finished from the user's perspective.
+	baseCtx withContext
+
+	mu struct {
+		syncutil.Mutex
+		// appRespSent and storageAppendRespSent track whether range messages have
+		// already been logged to log at most once. This limits the log from growing
+		// too large at a small risk of missing some messages in the case of dropped
+		// messages or reproposals.
+		appRespSent, storageAppendRespSent map[raftpb.PeerID]bool
+	}
+}
+
+// logf logs the message to the trace context and the base context. The base
+// context is populated on the leaseholder and is attached to the SQL trace.
+func (tv *traceValue) logf(depth int, format string, args ...interface{}) {
+	log.InfofDepth(tv.ctx, depth+1, format, args...)
+	if tv.baseCtx != nil {
+		if ctx := tv.baseCtx.Context(); ctx != nil {
+			log.VEventfDepth(ctx, depth+1, 3, format, args...)
+		}
+	}
+}
+
+// String attempts to balance uniqueness with readability by only keeping the
+// lower 16 bits of the trace and span.
+func (tv *traceValue) String() string {
+	return redact.StringWithoutMarkers(tv)
+}
+
+func (tv *traceValue) SafeFormat(w redact.SafePrinter, _ rune) {
+	w.Printf("i%d/%x.%x", tv.traced.Index, uint16(tv.traced.TraceID), uint16(tv.traced.SpanID))
+}
+
+// RaftTracer is a utility to trace the lifetime of raft log entries. It may
+// include some unrelated entries, since it does not consider the term. It
+// traces at most one MsgAppResp and MsgStorageAppendResp per index which is the
+// first one that is past our index entry. This limitation means it may not
+// capture all the relevant messages particularly if the term changes.
+//
+// The library will log in two different ways once to the standard cockroach log
+// and once to the SQL trace on the leader.
+// TODO(baptist): Look at logging traces on followers as well and sending back
+// to the leader in the MsgAppResp. It would need to be best effort, but might
+// still be useful.
+type RaftTracer struct {
+	// m is a map of all the currently traced entries for this replica. The
+	// aggregate size of the map is equal to or less than numRegistered. We add
+	// to numRegistered before we update m, and delete from m before we remove
+	// from numRegistered to keep this invariant.
+	m syncutil.Map[kvpb.RaftIndex, traceValue]
+	// numRegistered is the number of currently registered traces for this
+	// store, not this replica. The number of registered will always be less
+	// than the MaxConcurrentRaftTraces setting. If the setting is lowered, we
+	// flush all traces on all replicas.
```

Comment:
Same thing: we flush lazily, so it's not always the fact that the limit is not exceeded.

Reply:
Ack